main/rest/leaf.py

```diff
@@ -229,24 +238,107 @@ def _get(self, params):
     def _patch(self, params):
         obj = Leaf.objects.get(pk=params['id'], deleted=False)
         model_dict = obj.model_dict
+        grandparent = obj.parent
 
         # Patch common attributes.
         if 'name' in params:
             obj.name = params['name']
-            obj.save()
+        
         new_attrs = validate_attributes(params, obj)
         obj = patch_attributes(new_attrs, obj)
+        obj.save()
+
+        # Change the parent of a leaf
+        if 'parent' in params:
+            if params['parent'] != params['id']:
+                children = Leaf.objects.filter(parent=obj.id, deleted=False)
+                if params['parent'] is None or params['parent'] == -1:
+                    obj.parent = None
+                else:
+                    obj.parent = Leaf.objects.get(pk=params['parent'], deleted=False)
+                obj.path = obj.computePath()
+                obj.save()
+
+                #Update child path, or parent
+                newparent = params['parent']
+                self._update_children_newparent(self, children, grandparent, newparent)
+            
+
         obj.save()
         log_changes(obj, model_dict, obj.project, self.request.user)
-        return {'message': 'Leaf {params["id"]} successfully updated!'}
+        return {'message': f'Leaf {params["id"]} successfully updated!'}
 
     def _delete(self, params):
         leaf = Leaf.objects.get(pk=params['id'], deleted=False)
         project = leaf.project
-        model_dict = leaf.model_dict
-        delete_and_log_changes(leaf, project, self.request.user)
-        TatorSearch().delete_document(leaf)
-        return {'message': 'Leaf {params["id"]} successfully deleted!'}
+
+        ids = self._get_children_id_set(self, params['id'])
+
+        queryset = Leaf.objects.filter(pk__in=ids)
+        for i in ids:
+            inner_leaf = Leaf.objects.get(pk=i, deleted=False)
+            TatorSearch().delete_document(inner_leaf)
+                
+        bulk_delete_and_log_changes(queryset, project, self.request.user)
+
+        # todo figure out syntax for this query
+        # query = get_leaf_es_query(params)
+        # TatorSearch().delete(project, query)
+
+       
+        if len(ids) == 2 :
+            return {'message': f'Leaf {params["id"]} and {len(ids) - 1} child successfully deleted! '}
+        elif len(ids) > 2 :
+            return {'message': f'Leaf {params["id"]} and {len(ids) - 1} children successfully deleted! '}
+        else:
+             return {'message': f'Leaf {params["id"]} successfully deleted! '}
+
+    def _get_children_id_set(self, leaf_id):
+        ch_list = self._recursive_inner_child(self, leaf_id, [])
```

Comment:
Ah, now you don't need `self` in the argument list to this function
```suggestion
        ch_list = self._recursive_inner_child(leaf_id, [])
```